File: RTS Project/gen_task_file.py
import random

def uunifast(total_utilization, num_tasks):
    # Generate random utilizations
    utilizations = []
    sumU = total_utilization
    for i in range(1, num_tasks):
        next_sumU = sumU * random.random() ** (1 / (num_tasks - i))
        utilizations.append(sumU - next_sumU)
        sumU = next_sumU
    utilizations.append(sumU)

    return utilizations

def generate_taskset(num_tasks, total_utilization, max_period):
    taskset = []
    utilizations = uunifast(total_utilization, num_tasks)
    for i in range(num_tasks):
        period = random.randint(1, max_period)
        execution_time = random.randint(1, period) # Random execution time <= period
        deadline = random.randint(execution_time, period) # Random deadline between execution time and period
        taskset.append((i+1, f"Task {i+1}", utilizations[i], period, execution_time, deadline))
    return taskset

def save_taskset_to_file(taskset, filename):
    with open(filename, 'w') as file:
        file.write(f"{len(taskset)}\n")
        for task in taskset:
            file.write(f"{task[0]} {task[1]} {task[2]} {task[3]} {task[4]} {task[5]}\n")

if __name__ == "__main__":
    num_tasks = 10
    total_utilization = 0.8
    max_period = 100
    taskset = generate_taskset(num_tasks, total_utilization, max_period)
    save_taskset_to_file(taskset, "task_input.txt")

File: RTS Project/gen_task_file.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Task structure representing tasks in a flight cockpit
typedef struct {
    int id;          // Task ID
    char operation[100]; // Description of the task operation
    int criticality; // Criticality level (1 to 5)
    int execution_time; // Execution time (in seconds)
    int deadline;    // Deadline (in seconds)
} Task;

// Function to generate a random task
Task generate_random_task(int id) {
    Task task;
    char *operations[] = {"Sensor Reading: Altitude", "Control Input: Throttle Adjustment",
                          "System Check: Fuel Level", "Warning: Airspeed Low",
                          "Navigation Update: Waypoint Adjustment"};
    int num_operations = sizeof(operations) / sizeof(operations[0]);
    
    task.id = id;
    sprintf(task.operation, "%s", operations[rand() % num_operations]);
    task.criticality = rand() % 5 + 1;
    task.execution_time = rand() % 5 + 1;
    task.deadline = rand() % 15 + 1;
    
    return task;
}

int main() {
    srand(time(NULL)); // Seed the random number generator

    int num_tasks = rand() % 20 + 1; // Generate a random number of tasks (1 to 20)

    FILE *file = fopen("output_tasks.txt", "w");
    if (file == NULL) {
        printf("Error opening file.\n");
        return 1;
    }

    // Write number of tasks to file
    fprintf(file, "%d\n", num_tasks);

    // Generate and write tasks to file
    for (int i = 0; i < num_tasks; i++) {
        Task task = generate_random_task(i + 1);
        fprintf(file, "%d %s %d %d %d\n", task.id, task.operation, task.criticality, task.execution_time, task.deadline);
    }

    fclose(file);
    printf("Tasks generated and written to output_tasks.txt.\n");

    return 0;
}
